### ISSUE TITLE
Exclude API paths in CSharp Client generation

### ISSUE THREAD

**Issue by SaebAmini** [RicoSuter/NSwag#2584]
I know there's already support for excluding controllers when generating the OpenAPI document, but this is for the other end when generating clients.

Is it possible to configure NSwag to generate clients only for specific segments/paths from a given OpenAPI document? I'm specifically using the CSharp Client generator here.

The use case is for when you're dealing with huge APIs, but you only really need to consume small portions of it.

**Comment by jeremyVignelles** [comment]
I had a similar issue where I have a "Common" project with its API, and my project that adds its own API. When using swagger on the web, I want to be able to see both, but when generating the project's client classes, I only want to generate that portion of the API.

I found this workaround:

In my AddOpenApiDocument() :
```cs
            if (Assembly.GetEntryAssembly() != Assembly.GetExecutingAssembly())
            {// This is true when the API is being generated with NSwag
                settings.ApiGroupNames = new[] {"Project"};
            }
```

I hope that it helps.

**Comment by RicoSuter** [comment]
@jeremyVignelles that's on the OpenAPI generator side, but the question here is whether there is a filter setting for the incoming OpenAPI spec in the client code generator, and the answer is: Not yet.

So the solution for now is to preprocess/transform the spec (e.g. with a simple cmd line app which uses the nswag packages) before passing it to the generator.

**Comment by jeremyVignelles** [comment]
Yes, maybe that wasn't clear, but what I meant is that I'm using NSwagStudio to generate the client code from the .net assembly. In that process, the entry assembly is NSwag, which in turn runs the code assembly.

That way, I'm able to generate C#/TS code from a subset of the API.

**Comment by RicoSuter** [comment]
I agree, if you control the whole toolchain/pipeline you can generate a spec with only the operations you need. But if you get the spec from an external source, then you dont have the first step and you need to apply a filter on the document.

**Comment by SaebAmini** [comment]
Thanks for the suggestion @jeremyVignelles. As @RicoSuter is saying, this is for an external API document that I can't modify.

@RicoSuter do you think this is something I could pick up and do a PR for? If you could point me to the correct places I'll have a crack at it.

**Comment by RicoSuter** [comment]
The hardest part here is that for example choose only half of the operations and we only need to render the required DTO classes. That means we need to transitively scan the schemas and only pick the needed ones depending on their usage.

**Comment by unchase** [comment]
In fact, you only need to preload the DTO classes from the specification in NSwagStudio, deduce all the classes, and then select only the necessary ones. Then generate client code.
Is it possible to implement this?

And then you can think about how to add this to the [NSwag.Commands](https://www.nuget.org/packages/NSwag.Commands/) - this is the hardest part. Right?

**Comment by francisminu** [comment]
I am generating the spec using the tool StopLightStudio.We currently pass this spec to NSwag dll to get the Controllers and the Models generated. But, going forward we would like to have just the Models genert and not the controller. Is the ignoreApi attribute used for this purpose? If so, where do I have to specify this attribute?

Please find my sample spec below:

`openapi: 3.0.0
info:
  title: PhysicalInventory
  version: '1.0'
servers:
  - url: 'http://localhost:3000'
paths:
  /tasks/create:
    put:
      summary: ''
      operationId: put-tasks-create
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/CreateTaskResponseModel'
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/CreateTaskRequestModel'
    options:
      summary: ''
      operationId: options-tasks-create
components:
  schemas:
    CreateTaskRequestModel:
      title: CreateTaskRequestModel
      allOf:
        - $ref: ../shared/BaseRequestModel.v1.yaml
        - type: object
          properties:
            TaskId:
              type: string
            List:
              type: array
              items:
                type: string
            Task:
              $ref: '#/components/schemas/TaskModel'
            TaskList:
              type: array
              items:
                $ref: '#/components/schemas/TaskModel'
    CreateTaskResponseModel:
      allOf:
        - $ref: ../shared/BaseResponseModel.v1.yaml
        - type: object
          properties:
            IsError:
              type: string
    TaskModel:
      title: TaskModel
      type: object
      properties:
        id:
          type: string
`

**Comment by dis081278** [comment]
Hi, I'm using NSwag Studio to generate Clients via processing of the Web API by reflection.  There are actions  [such as HttpGet, Route("")] in my controllers for which I wish to avoid generating client methods.  Does a process exist yet where I can specify which actions to skip using the Web API by reflection method?

**Comment by RicoSuter** [comment]
Use an [OpenApiIgnore] attribute or preprocess the spec before providing it to the client generator. With the ASP.NET Core spec generator you can also register custom operation processors to exclude operations or define other ApiGroupNames as @jeremyVignelles described.

**Comment by dis081278** [comment]
Thanks.  That's what I was looking for!

**Comment by tiagopsantos** [comment]
This is really tricky, and a common problem
Huge json APIs, but only a small amount needed 

An ignoreApi or an onlyApi would be helpful

Other alternative is to edit manually the json API... and remove the unnecessary

**Comment by Sleyy** [comment]
Are there any updates on the topic? 
We also have a huge api being generated which I would like to filter by tag name for example and ignore the rest.

**Comment by dsolteszopyn** [comment]
"documentGenerator": {
    "fromDocument": {
      "json": "",
      "url": "",
      "output": null,
      "newLineBehavior": "Auto",
      "operationFilter": {
        "includeOperations": ["operationId1", "operationId2"],
        "excludeOtherOperations": true
      }
    }
  }
  
  need to have some type of operationfilter that supports either includeOperations or excludeOperations

**Comment by bdovaz** [comment]
In my case I solved it with:
* A powershell script that parses the original OpenAPI json, removes the paths I don't need and writes a different OpenAPI json.
* Create and additional MSBuild task that runs the mentioned powershell script and another task with the nswag commands "as usual" but using this transformed OpenAPI json.

I had to do it, not only to be able to generate a lighter client with the endpoints that I only need, but also because NSwag does not support the OpenAPI specification correctly, it generated incorrectly (it does not compile the code) certain endpoints that I don't even use.

**Comment by dimyle** [comment]
I've created a dotnet tool for OpenApi files reducing/filtering [https://www.nuget.org/packages/Deimdal.ExtractOpenApi](see this nuget)

**Comment by molinch** [comment]
> [@jeremyVignelles](https://github.com/jeremyVignelles) that's on the OpenAPI generator side, but the question here is whether there is a filter setting for the incoming OpenAPI spec in the client code generator, and the answer is: Not yet.
> 
> So the solution for now is to preprocess/transform the spec (e.g. with a simple cmd line app which uses the nswag packages) before passing it to the generator.

@RicoSuter I guess this still holds, right?
